sphinxcontrib/httpdomain.py

```diff
@@ -692,6 +692,17 @@ def get_objects(self):
                 anchor = http_resource_anchor(method, path)
                 yield (path, path, method, info[0], anchor, 1)
 
+    def merge_domaindata(self, docnames, otherdata):
+        """Merge domaindata from the workers/chunks when they return.
+
+        Called once per parallelization chunk.
+        Only used when sphinx is run in parallel mode.
+
+        :param docnames: a Set of the docnames that are part of the current chunk to merge
+        :param otherdata: the partial data calculated by the current chunk
+        """
+        for x in ('options', 'head', 'post', 'get', 'put', 'patch', 'delete', 'trace', 'connect', 'copy', 'any'):
+            self.data[x].update(otherdata[x])
```

Comment:
When I wanted to patch this in my local build environment, I had this for `merge_domaindata` instead, but I'm not sure which is more effective, yours or mine:
``` python
def merge_domaindata(self, docnames, otherdata):
    for typ in self.object_types:
        for key, data in otherdata[typ].items():
            if data[0] in docnames:
                self.data[typ][key] = data
```

Reply:
Hi,
on my test data with "-j 8" summing up the execution times of merge_domaindata, execution times of both solutions are not really distinguishable, however if I omit the "if data[0] in docnames" line of your solution (which also makes both solutions semantically equivalent) then yours has a slight advantage.
I'm fine with either of the approaches (of course using self.object_types instead of repeating the HTTP verbs is preferrable) though personally I'd prefer the update()-call.

Is there any particular reason why you included the if-line ? Could it be that this always evaluates to True, or is it my test-data being too simple ?

Reply:
Thanks for your investigation and explanation, @hofmandl1! I cobbled together my patch by close reading of several implementations of `merge_domaindata` in the Sphinx core distribution domains, so I honestly cannot say that I fully understand all the data structures that get passed around or what they contain.

What I assumed was that if you have a batch of otherdata, and a list of docnames, then you probably don't want to add an entry to `self` from the otherdata unless `data[0]` (which I understood was a docname) also exists in `self's` list of docnames (otherwise you are merging in domain data that doesn't belong in self).

This assumption may be entirely incorrect.

I'm completely happy for someone with more experience to guide to the best solution, and then I'll just fix and/or remove my own implementation.

Reply:
I also like your use of `update` as well, but it seems to privilege the entry in `self` over the entry that might exist in `otherdata`... I'm not sure which would be better (I think my own implementation always overwrite any entry already in `self` while yours never will, and I'm unsure which is correct... probably yours?)

Reply:
Hi, my experience/knowledge with the merge_domaindata is not much different .. I also googled for some existing implementations :-$ .. did some debugging prints and checked the outcome ..
According to this https://docs.python.org/3/library/stdtypes.html#dict.update existing values would be overwritten by update just like in your solution

Reply:
Ah! Yes, I was confusing it with `setdefault`. In that case, I definitely prefer your `update()` sol'n. 8)

Reply:
Hmm .. in any case, a proper implementation probably should emit a sphinx warning if the same entrypoint is defined on several pages. I wouldn't know how to do that. I see, that in other methods, `reporter.warning()` is invoked, however it seems we don't have an "env"-instance available like we have in `resolve_xref` which is used to get a `reporter`-instance. Also when looking for this I stumbled on the comment in line 30:
```
# The env.get_doctree() lookup results in a pickle.load() call which is
# expensive enough to dominate the runtime entirely when the number of endpoints
# and references is large enough. The doctrees are generated during the read-
# phase and we can cache their lookup during the write-phase significantly
# improving performance.
# Currently sphinxcontrib-httpdomain does not declare to support parallel read
# support (parallel_read_safe is the default False) so we can simply use a
# module global to hold the cache.
_doctree_cache = {}
```
so maybe this plugin is not declared parallel-read-safe for a good reason.

For me the solution works, but without further help from someone with more sphinx experience this probably is not mature enough for master. I guess this is not as easy as I thought :(

Reply:
Seems I found how it could work: https://github.com/sphinx-doc/sphinx/pull/4882/commits/f4f693eff7b081785cc5704a6ff22cf0371355f3

Reply:
The warning now seems to be emitted correctly and also to abort the build if sphinx is e.g. called via the "-W" option. However I'm unsure about the __doctree_cache thing ..
Does this really prevent the module from being declared parallel-read-safe ?

Reply:
Without depending on the cache but directly loading `env.doctree(fromdocname)` every time most definitely slows down the write stage. I believe that the write stage doesn't happen until all workers complete reading, so it seems to me like the cache wouldn't ever get stuck with a node in it that's only partially read?

Looking at the way `resolve_xref` seems to be written in the domains in Sphinx core, the one here seems potentially overcomplicated.

It seems like the only reason the code has for wanting to make a call on `env.get_doctree(fromdocname)` is so that the code has access to the `reporter` attribute on the document for error reporting.

If the Domain could get a general reporter without having to load the entire doctree of a fromdocname each time, then it wouldn't need to cache or load the fromdocname doctree at all, perhaps? Looks like some refactoring might be useful here.

Reply:
I've figured out a way to get rid of the doctree cache and it seems to be working with my documentation suite. It has these points:

1. Use a `SimpleNamespace` from the `types` standard library:
```python
from types import SimpleNamespace
```

2. Create a "shadow document" object in the Domain who's only purpose is to hold a LoggingReporter that can be passed around:
```python
# if from docname not in _doctree_cache:
#   _doctree_cache[fromdocname] = env.get_doctree(fromdocname)
# doctree = _doctree_cache[fromdocname]

# resnode = role.result_nodes(doctree, env, node, None)[0][0]

shadow_doc = SimpleNamespace(
    reporter=LoggingReporter(env.doc2path(fromdocname)))
resnode = role.result_nodes(shadow_doc, env, node, None)[0][0]

...
     # reporter = doctree.reporter
     # reporter.warning('Cannot resolve reference to %r' % text,
     shadow_doc.reporter.warning('Cannot resolve reference to %r' % text,
```

3. Ensure that HTTPXRefRole implements `result_nodes` to ensure that getting a shadow_document is not going to get up into its super():
```python
def result_nodes(self, document, env, node is_ref):
   return [node], []
```

This copies what its super() currently does, but puts a firewall in place in case that changes.

This lightens the load on getting ahold of the reporter for a document, without having to load in its entire doctree.

Not sure if you want to think about putting this into this PR, or into another one potentially.

Reply:
hmm .. so you don't like the idea of upping the sphinx-requirement to 1.6? the logger also supports location specification, see https://www.sphinx-doc.org/en/master/extdev/logging.html#sphinx.util.logging.SphinxLoggerAdapter

Also the changelog for 1.6b1 says:
```
#2367: Sphinx.warn(), Sphinx.info() and other logging methods are now deprecated. Please use sphinx.util.logging (Logging API) instead.
```
and newer sphinx modules all seem to have this global `logger` field.

This way I think we wouldn't need the NameSpace-Hack ..

Reply:
I don't mind upping the requirements to 1.6 at all, personally, and I didn't know about using a global logger; I'd be totally happy to see the removal of passing around a document within the Domain and it's XRef nodes, and replaced with using the global logger for messaging.

Reply:
ah, but I hadn't seen the many implementations of result_nodes .. ok .. yep .. I'll use the namespace proposal in order not to touch too many code

Reply:
nope .. types.SimpleNamespace is only available since python 3.3

Reply:
That's disappointing. 8( One could accomplish the same with a thin, internal "shadow_document" class... using the global logger and just not passing down a document at all is looking better and better... I liked the SimpleNamespace/internal class idea because it helped localize changes to just the Domain and leave the XRef nodes nicely generic.